Remove {f,F}orall_goto_functions

tautschnig opened the pull request:
This was useful in the past, but with C++-11 we can use a ranged-for to
avoid the iterator altogether (in all but the cases in goto-diff, where
we now make the iterator explicit).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented:
Yes please. Though I think it’d probably be a bit too ambitious to do them all in one PR

tautschnig committed:
Remove {f,F}orall_goto_functions

This was useful in the past, but with C++-11 we can use a ranged-for to
avoid the iterator altogether (in all but the cases in goto-diff, where
we now make the iterator explicit).

martin-cs commented on src/analyses/flow_insensitive_analysis.h:
```diff
@@ -173,7 +173,7 @@ class flow_insensitive_analysis_baset
 
   bool fixedpoint(
-    goto_functionst::function_mapt::const_iterator it,
+    const goto_functionst::function_mapt::value_type &gf_entry,
```
This feels like it might be better as a `goto_programt`?

martin-cs commented on src/analyses/flow_insensitive_analysis.cpp:
```diff
@@ -383,20 +384,21 @@ void flow_insensitive_analysis_baset::fixedpoint(
 
 bool flow_insensitive_analysis_baset::fixedpoint(
-  const goto_functionst::function_mapt::const_iterator it,
+  const goto_functionst::function_mapt::value_type &gf_entry,
```
`goto_programt` ?

martin-cs commented on src/goto-checker/symex_coverage.cpp:
```diff
@@ -52,7 +52,7 @@ class goto_program_coverage_recordt : public coverage_recordt
   goto_program_coverage_recordt(
     const namespacet &ns,
-    goto_functionst::function_mapt::const_iterator gf_it,
+    const goto_functionst::function_mapt::value_type &gf_entry,
```
Feels a little cumbersome.  Isn't this a `goto_functiont`?

tautschnig replied:
These also require the name of the `goto_functiont`. What could be done is passing that information as two arguments? That would probably make for are clearer interface. Let me fix this.

tautschnig replied:
Now done.

martin-cs reviewed (approved):
Thanks for these patches; sorry I missed this one.
